Engine/DataFeeds/Enumerators/SynchronizingEnumerator.cs

```diff
@@ -26,18 +26,19 @@ namespace QuantConnect.Lean.Engine.DataFeeds.Enumerators
     /// Represents an enumerator capable of synchronizing other base data enumerators in time.
     /// This assumes that all enumerators have data time stamped in the same time zone
     /// </summary>
-    public class SynchronizingEnumerator : IEnumerator<BaseData>
+    public abstract class SynchronizingEnumerator<T> : IEnumerator<T>
     {
-        private IEnumerator<BaseData> _syncer;
-        private readonly IEnumerator<BaseData>[] _enumerators;
+        private IEnumerator<T> _syncer;
+        private readonly IEnumerator<T>[] _enumerators;
+        protected abstract DateTime GetInstanceTime<T>(T instance);
```

Comment:
- Let's add the docs for `GetInstanceTime` 👍 
- Constructors can now be protected too